tensorflow/compiler/xla/service/all_reduce_reassociate.cc

```diff
@@ -22,24 +22,88 @@ limitations under the License.
 #include "tensorflow/compiler/xla/service/collective_ops_utils.h"
 #include "tensorflow/compiler/xla/service/hlo_domain_map.h"
 #include "tensorflow/compiler/xla/service/hlo_query.h"
+#include "tensorflow/compiler/xla/service/pattern_matcher.h"
 #include "tensorflow/tsl/platform/errors.h"
 
 namespace xla {
 namespace {
 
+namespace m = match;
+
+bool AreAllreduceKeysEqual(AllReduceKey &key0, AllReduceKey &key1,
+                           bool ignore_element_type) {
+  // We compare all elements in the tuple except element_type which is the
+  // second member in the AllReduceKey tuple.
+  if (ignore_element_type) {
+    return std::get<0>(key0) == std::get<0>(key1) &&
+           std::get<2>(key0) == std::get<2>(key1) &&
+           std::get<3>(key0) == std::get<3>(key1) &&
+           std::get<4>(key0) == std::get<4>(key1) &&
+           std::get<5>(key0) == std::get<5>(key1);
+  } else {
+    return key0 == key1;
+  }
+}
 // Returns if the given all reduce instructions are compatible with each other.
 // Note that since the given all-reduce instructions are connected to another
 // instruction by a direct data flow edge, they must belong to the same domain.
 // As a result, we don't need to include any domain information in the
 // AllReduceKey to check compatibility.
 bool AreCompatible(const HloAllReduceInstruction *ar0,
-                   const HloAllReduceInstruction *ar1, ReductionKind op_kind) {
+                   const HloAllReduceInstruction *ar1, ReductionKind op_kind,
+                   bool ignore_element_type) {
   std::optional<AllReduceKey> key0 = GetAllReduceKey(ar0);
   std::optional<AllReduceKey> key1 = GetAllReduceKey(ar1);
   auto kind0 = MatchReductionComputation(ar0->to_apply());
-  return key0 && key1 && kind0 && *key0 == *key1 && kind0 == op_kind;
+  // If ignore_element_type is true, then we compare all elements in the
+  // AllReduceKey tuple except the element_type
+  return key0 && key1 && kind0 &&
+         AreAllreduceKeysEqual(*key0, *key1, ignore_element_type) &&
+         kind0 == op_kind;
+}
+
+bool AreCompatibleConverts(const HloInstruction *convert0,
+                           const HloInstruction *convert1) {
+  bool is_compatible = true;
+  // For numerical stability, we only re-order ops with casts from a narrow type
+  // to a wider type.
+  if (convert0) {
+    is_compatible &= primitive_util::CastPreservesValues(
+        convert0->operand(0)->shape().element_type(),
+        convert0->shape().element_type());
+  }
+
+  if (convert1) {
+    is_compatible &= primitive_util::CastPreservesValues(
+        convert1->operand(0)->shape().element_type(),
+        convert1->shape().element_type());
+  }
+
+  if (convert0 && convert1) {
+    CHECK(convert0->operand(0)->shape().element_type() ==
```

Comment:
I don't think this is correct. What I meant was check that convert0->shape() == convert1->shape() and then check compatibility by checking operand element types. I think your earlier change did both and I just commented that one of them is redundant.

Reply:
ah i see, misunderstood your comment earlier. will change it.